Update install instructions, simplify requirements

arve0 opened the pull request:
This tries to simplify the requirements and also updates install instructions. It will most probably not work out of the box. Fixes #1506 (when all errors are ironed out).

I'm open to suggestions!

blink1073 commented on setup.py:
```diff
@@ -105,9 +86,10 @@ def configuration(parent_package='', top_path=None):
         configuration=configuration,
-        setup_requires=SETUP_REQUIRES,
         install_requires=INSTALL_REQUIRES,
-        requires=REQUIRES,
```
This is used to populate the PyPI entry, which does not pick up `install_requires`.

arve0 replied:
I'll put it back. Should it hold both runtime and compile dependencies?

blink1073 replied:
I'd say just runtime.

blink1073 replied:
But please also add a comment since apparently this was a point of confusion.  :smile:

arve0 replied:
It was already commented, but I was to tired to catch it!

blink1073 commented on tools/travis_before_install.sh:
```diff
@@ -27,20 +27,24 @@ retry () {
+     $TRAVIS_PYTHON_VERSION == 3.3 -o \
+     $TRAVIS_PYTHON_VERSION == 3.4 ]; then
+# use conda env
+echo jej
```
What is this?

arve0 replied:
just me trying out bash if, it will be removed

blink1073 commented on tools/travis_before_install.sh:
```diff
@@ -27,20 +27,24 @@ retry () {
-if [[ $TRAVIS_PYTHON_VERSION == 2.7* ]]; then
-    virtualenv --system-site-packages ~/venv
-    sudo apt-get install python-scipy python-matplotlib python-imaging
-    sed -i 's/cython>=/cython==/g' requirements.txt
```
We specifically wanted to test against the minimum versions of Cython and NetworkX here to make sure we weren't introducing new features that were incompatible.

arve0 replied:
OK. Shall we test all minimum requirements? Not only networkx and cython?

blink1073 replied:
Sounds good to me, you could just replace `>=` with `==` globally.

arve0 commented:
@blink1073 It's a bit havoc in the code at the time, I'll clean it up when things are starting to look good. Are the `numpy.distutils.core.setup` required for cython builds?

blink1073 commented:
Correct.

arve0 commented:
Ok, then I will add instructions for those getting scikit-image as source package, and fail more gracefully with instruction for installing scipy, numpy and cython.

arve0 commented:
Anyone have an idea why the appveyor build does not seem to link cython files correctly?

blink1073 commented:
You took Cython out of requirements.txt, try adding a specific `build` install on this [line](https://github.com/scikit-image/scikit-image/blob/master/appveyor.yml#L42)

blink1073 commented on tools/travis_before_install.sh:
```diff
@@ -27,20 +27,24 @@ retry () {
-    sed -i 's/networkx>=/networkx==/g' requirements.txt
-    sed -i '/pillow/d' requirements.txt
-else
-    virtualenv -p python --system-site-packages ~/venv
```
This was needed to make the plot examples work.  Right now they're timing out in the 2.7 build.

arve0 replied:
Do you know what was the magic sauce, venv, pillow, scipy or matplotlib?

System pillow and scipy are not meeting the requirements, should we lower the requirements? If the tests pass, I guess this means we require pillow>=1.1.7 and scipy>=0.9.0.

blink1073 replied:
There's something in the system site packages (maybe Tk) that was needed to make the matplotlib `Template` backend work without hanging.

blink1073 replied:
Agreed on the version changes.

arve0 commented on appveyor.yml:
```diff
@@ -37,7 +37,7 @@ install:
 
   # Install the build and runtime dependencies of the project.
-  - "%CMD_IN_ENV% pip install -v %WHEELHOUSE% -r tools/appveyor/requirements.txt"
+  - "%CMD_IN_ENV% pip install --trusted-host 28daf2247a33ed269873-7b1aad3fab3cc330e1fd9d109892382a.r6.cf2.rackcdn.com -r tools/appveyor/requirements.txt"
```
@blink1073 This is the only change to the appveyor env. The log shows that cython is installed from the wheel house and that files are compiled with cython.

blink1073 replied:
Maybe try taking this addition out?  Otherwise, it must have to do with changes made in `setup.py` or `requirements.txt` (but I'm confused at this point).

arve0 replied:
Any reason why both wheel and `bdist_wininst` is built/packaged?

blink1073 replied:
I don't think so, wheels are the way forward. 



—
Sent from Mailbox

On Sun, May 24, 2015 at 6:14 AM, Arve Seljebu <notifications@github.com>
wrote:

>> @@ -37,7 +37,7 @@ install:
>>    - "python -c \"import struct; print(struct.calcsize('P') * 8)\""
>>  
>>    # Install the build and runtime dependencies of the project.
>> -  - "%CMD_IN_ENV% pip install -v %WHEELHOUSE% -r tools/appveyor/requirements.txt"
>> +  - "%CMD_IN_ENV% pip install --trusted-host 28daf2247a33ed269873-7b1aad3fab3cc330e1fd9d109892382a.r6.cf2.rackcdn.com -r tools/appveyor/requirements.txt"
> Any reason why both wheel and `bdist_wininst` is built/packaged?
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/scikit-image/scikit-image/pull/1526/files#r30951835

arve0 commented:
Ok, so this are the purposed changes. (I'll clean up the commit noise and Silvester's suggestions later). Note that I'm talking about fresh installs here, neither having runtime or build requirements met.

### Changes in skimage
1. Now requires scipy, numpy and dask[array].
2. Removed some magic in setup.py.
3. Gracefully fail when running setup.py and missing numpy.

The effect is that when scikit-image is installed from binaries (read wheel, that is OS X as of now) and setup.py is not ran, dependencies will be correctly installed.

For the source package, it's not the same story, as it will always run setup.py (and fail), never getting to `install_requires`. Though, if pip are able to run setup.py, it can resolve runtime dependencies (which are overlapping with build dependencies) and build dependencies (setup_requires, a setuptools easy_install directive). So we put build dependencies which do not overlap with runtime dependencies (cython), in `setup_requires`.

### Changes in both test environments
1. pip now need `--trusted-host` to install wheels from http

### Changes in travis
1. Test minimum requirements
  - To not tamper with requirements.txt in other cases
2. Only install binaries from wheel house
  - Let setup.py/install_requires sort out the rest
3. Do not use venv (if I can get it to pass without)

### Changes to appveyor
1. Do not package bdist_wininst

arve0 commented:
Unable to find out when six.moves.urllib was added to six (its above 1.3). Found this: https://bitbucket.org/gutworth/six/pull-request/5/create-sixmovesurllib/diff

Anyone familiar with gitbucket?

arve0 commented:
Seems to be 1.4.0

blink1073 commented:
@arve0, the reason we want to use our own venv as opposed to the default Travis one is that it kept changing, and breaking our builds.  By rolling our own we insulate ourselves from these changes.

arve0 commented:
Aha. I saw that numpy was already installed in the travis env. I'll use venv for all python version then.

arve0 commented:
@jjhelmus Hi! I see you are the creator of the appveyor build, can you help me out? Any idea why the build and wheel install is succeeding but the cython import fails?

https://ci.appveyor.com/project/stefanv/scikit-image-1fxlb/build/1.0.366

Some excerpts:

**707**
```
Compiling C:\projects\scikit-image-1fxlb\skimage\_shared\geometry.pyx because it changed.
Cythonizing C:\projects\scikit-image-1fxlb\skimage\_shared\geometry.pyx
```

**794**
`building extension "skimage._shared.geometry" sources`

**1323**
```
building 'skimage._shared.geometry' extension
compiling C sources
creating build\temp.win32-2.7
creating build\temp.win32-2.7\Release
creating build\temp.win32-2.7\Release\skimage
creating build\temp.win32-2.7\Release\skimage\_shared
C:\Program Files (x86)\Common Files\Microsoft\Visual C++ for Python\9.0\VC\Bin\cl.exe /c /nologo /Ox /MD /W3 /GS- /DNDEBUG -IC:\Python27_32\lib\site-packages\numpy\core\include -IC:\Python27_32\include -IC:\Python27_32\PC /Tcskimage\_shared\geometry.c /Fobuild\temp.win32-2.7\Release\skimage\_shared\geometry.obj
Found executable C:\Program Files (x86)\Common Files\Microsoft\Visual C++ for Python\9.0\VC\Bin\cl.exe
C:\Python27_32\lib\site-packages\setuptools\dist.py:285: UserWarning: Normalizing '0.12dev' to '0.12.dev0'
  normalized_version,
C:\Program Files (x86)\Common Files\Microsoft\Visual C++ for Python\9.0\VC\Bin\link.exe /DLL /nologo /INCREMENTAL:NO /LIBPATH:C:\Python27_32\libs /LIBPATH:C:\Python27_32\PCbuild /EXPORT:initgeometry build\temp.win32-2.7\Release\skimage\_shared\geometry.obj /OUT:build\lib.win32-2.7\skimage\_shared\geometry.pyd /IMPLIB:build\temp.win32-2.7\Release\skimage\_shared\geometry.lib /MANIFESTFILE:build\temp.win32-2.7\Release\skimage\_shared\geometry.pyd.manifest
Found executable C:\Program Files (x86)\Common Files\Microsoft\Visual C++ for Python\9.0\VC\Bin\link.exe
mt.exe -nologo -manifest build\temp.win32-2.7\Release\skimage\_shared\geometry.pyd.manifest -outputresource:build\lib.win32-2.7\skimage\_shared\geometry.pyd;2
```

**1938**
`copying build\lib.win32-2.7\skimage\_shared\geometry.pyd -> build\bdist.win32\wheel\scikit_image-0.12.dev0.data\..\skimage\_shared`

**2247**
`Failure: ImportError (No module named _shared.geometry) ... ERROR`

**2336**
```
======================================================================
ERROR: Failure: ImportError (No module named _shared.geometry)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Python27_32\lib\site-packages\nose\loader.py", line 420, in loadTestsFromName
    addr.filename, addr.module)
  File "C:\Python27_32\lib\site-packages\nose\importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "C:\Python27_32\lib\site-packages\nose\importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "C:\Python27_32\lib\site-packages\skimage\draw\__init__.py", line 1, in <module>
    from .draw import circle, ellipse, set_color
  File "C:\Python27_32\lib\site-packages\skimage\draw\draw.py", line 3, in <module>
    from ._draw import _coords_inside_image
  File "_draw.pyx", line 1, in init skimage.draw._draw (skimage\draw\_draw.c:23754)
ImportError: No module named _shared.geometry
```

arve0 commented:
Seems like files in `build\temp.win32-2.7` are never copied to `build\bdist.win32\wheel`

arve0 commented:
Ah, found it at last! Cython is actually compiled from source at appveyor, not using version 20, as you've [already noted](https://github.com/scikit-image/scikit-image/pull/1526#issuecomment-104999687) @blink1073.

blink1073 commented:
Great progress so far @arve0, thanks for persevering.

arve0 commented:
Heh, what a mess! Think I've found the silver bullet. There have been some trouble with travis python 2.7 intervening with system python 2.7. To be specific, travis python have compiled `pkg_resources` of the system, giving me all this trouble.

arve0 commented:
Btw, do we want to test for PIL (python-imaging) also? If so, I will have to find up some cleverness for hindering the pillow wheel install.

blink1073 commented:
Yes, we explicitly support and tested against PIL. 



—
Sent from Mailbox

On Sun, May 24, 2015 at 5:18 PM, Arve Seljebu <notifications@github.com>
wrote:

> Btw, do we want to test for PIL (python-imaging) also? If so, I will have to find up some cleverness for hindering the pillow wheel install.
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/scikit-image/scikit-image/pull/1526#issuecomment-105071541

blink1073 commented on tools/travis_before_install.sh:
```diff
@@ -28,33 +28,49 @@ retry () {
+    sudo apt-get install python-setuptools
 else
-    virtualenv -p python --system-site-packages ~/venv
+    virtualenv -p $(which python) ~/venv
```
I think the `$(which python)` is what is breaking your build, try reverting this line.

arve0 commented on tools/travis_before_install.sh:
```diff
@@ -28,33 +28,49 @@ retry () {
+    # remove pillow from wheel install
+    export WHEELBINARIES=${WHEELBINARIES/pillow/}
+    # use system python
     virtualenv --system-site-packages ~/venv
```
@blink1073 I'm only running this 2.7 test of now, not using `-p` there. I [did get past](https://travis-ci.org/scikit-image/scikit-image/jobs/63893611) this `pkg_resource` error by installing `python-pip` and `python-setuptools` earlier on. But then it failed on `travis_script.sh`, trying to install the optional dependencies with pip.

Debugging the issue further showed that pip in `travis_script.sh` [was version 1.0](https://travis-ci.org/scikit-image/scikit-image/jobs/63893611#L3174), even if [pip was upgraded](https://travis-ci.org/scikit-image/scikit-image/jobs/63893611#L433) and [verified as version 7.0.1](https://travis-ci.org/scikit-image/scikit-image/jobs/63893611#L568).

So the issue that exists seem to be:
1. [System pkg_resources](https://travis-ci.org/scikit-image/scikit-image/jobs/63893611#L4981) is used, creating trouble for setuptools.
2. Something happens with the $PATH when jumping to the next script.

Also, here's the virtualenv manual on `-p`:
>  -p PYTHON_EXE, --python=PYTHON_EXE
>                        The Python interpreter to use, e.g.,
>                        --python=python2.5 will use the python2.5 interpreter
>                        to create the new environment.  The default is the
>                        interpreter that virtualenv was installed with
>                        (/usr/bin/python)

If I understand that correctly, one can be bitten if `virtualenv` was installed by another python intepreter. Not sure if that is possible, and we are not bitten by it: [travis 3.3 venv](https://travis-ci.org/scikit-image/scikit-image/jobs/63883429#L97), [our venv](https://travis-ci.org/scikit-image/scikit-image/jobs/63883429#L1245).

arve0 replied:
Adding: I think all the 2.7 trouble are due to travis venv [has python 2.7.9](https://travis-ci.org/scikit-image/scikit-image/jobs/63883427#L98), the virtualenv package in 2.7 seem to be installed by system python and [we get python 2.7.3 in our venv](https://travis-ci.org/scikit-image/scikit-image/jobs/63883427#L1431). Which is fair enough, as we are going to check system installed packages (scipy, matplotlib, etc).

A work around would of course be to use the wheels, with the drawbacks:
1. Testing deps with wheels, not guaranteeing scikit-image will work with minimum requirements from system packages. This might not be as bad as we think, as operating systems do ship pre-releases and report them as stable packages (ref llvm 3.5 on Ubuntu 14.04). A pre release should not be considered the same as the pypi stable release, and on can potentially stumble upon API-changes that break the build.
2. We will need to compile PIL 1.1.7 to wheel, compile it from source every time or drop the PIL dep.
3. Might be more that I'm unaware of...

arve0 commented:
I'm leaving it with this. I'm confident my changes are sound and what remains are test environment issues.

I've cleaned up the commit noise, the total should be:

### Changes in skimage
- Now requires scipy, numpy and dask[array].
  - The effect is that when scikit-image is installed from binaries (read *wheel*, that is OS X as of now), dependencies will be correctly installed. This is the only (consistent) way to resolve deps, as setup.py never runs when installing a wheel.
- Removed some magic in setup.py.
- Gracefully fail when running setup.py and missing numpy.

For the source installs, it's not the same story as with wheels, as it will run setup.py (and fail), never getting to install_requires. Though, if pip are able to run setup.py, it can resolve runtime dependencies (which are overlapping with build dependencies) and build dependencies (setup_requires, a setuptools easy_install directive). So we put build dependencies which do not overlap with runtime dependencies (cython), in setup_requires.

### Changes in travis
- Install binaries from wheel house first
  - Let pip requirements.txt sort out the rest

### Changes to appveyor
- Do not install cython 0.20
- resolve dask[array] from /requirements.txt

blink1073 commented:
@arve0, it looks like the latest appveyor build failed to find the requested version of Scipy in the wheelhouse.

arve0 commented:
Its due to changes in pip:
>  The repository located at 28daf2247a33ed269873-7b1aad3fab3cc330e1fd9d109892382a.r6.cf2.rackcdn.com is not a trusted or secure host and is being ignored. If this repository is available via HTTPS it is recommended to use HTTPS instead, otherwise you may silence this warning and allow it anyways with '--trusted-host 28daf2247a33ed269873-7b1aad3fab3cc330e1fd9d109892382a.r6.cf2.rackcdn.com'.

blink1073 commented:
Isn't this fun @arve0?  I remember encountering this exact problem before, but I cannot for the life of my figure out why your changes would cause it to return.

blink1073 commented:
It could have something to do with `six`.  We had previously installed it from the wheelhouse.

arve0 commented:
Define fun :joy:

@blink1073 I've come to the same conclusion, that's the only difference I have not tested out.

Travis [do warn](http://docs.travis-ci.com/user/languages/python/) about using system packages:
> CI Environment uses separate virtualenv instances for each Python version. System Python is not used and should not be relied on. If you need to install Python packages, do it via pip and not apt.

Its probably doable, but I do not think its worth my time.

I suggest one goes with *travis* python and add PIL to the wheel house to get a more stable testing system. Testing for Ubuntu LTS 12.04 *on travis* is kind of an corner case, getting the minimal dependencies correct should be enough. Also, I'm quite sure this isn't a *real world* issue, and even if it is the solution is obvious; Ubuntu LTS 12.04 users should compile the minimum requirements.

Same goes for appveyor - we should add cython wheel to the mix, reducing the test time by a lot.

blink1073 commented:
Agreed, I will handle the Linux wheels tomorrow. I say create another issue for the appveyor cython wheel. 



—
Sent from Mailbox

On Mon, May 25, 2015 at 10:42 PM, Arve Seljebu <notifications@github.com>
wrote:

> Define fun :joy:
> @blink1073 I've come to the same conclusion, that's the only difference I have not tested out.
> Travis [do warn](http://docs.travis-ci.com/user/languages/python/) about using system packages:
>> CI Environment uses separate virtualenv instances for each Python version. System Python is not used and should not be relied on. If you need to install Python packages, do it via pip and not apt.
> Its probably doable, but I do not think its worth my time.
> I suggest one goes with *travis* python and add PIL to the wheel house to get a more stable testing system. Testing for Ubuntu LTS 12.04 *on travis* is kind of an corner case, getting the minimal dependencies correct should be enough. Also, I'm quite sure this isn't a *real world* issue, and even if it is the solution is obvious; Ubuntu LTS 12.04 users should compile the minimum requirements.
> Same goes for appveyor - we should add cython wheel to the mix, reducing the test time by a lot.
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/scikit-image/scikit-image/pull/1526#issuecomment-105380300

jjhelmus commented:
Sorry just back from the weekend. Is there still an issue with the appveyor?  From the last run it looks like the only issue is that Cython is being built from source which is going to slow down the CI.  I can upload the [Cython 0.22 wheel files from Christoph Gohlke](http://www.lfd.uci.edu/~gohlke/pythonlibs/#cython) to the rackspace wheelhouse if that helps.   Another option is to move all of these Windows wheels to [binstar](https://binstar.org/about) which can be accessed using pip by pointing to https//pypi.binstar.org/username.  Continuum gives 3 GB of space for free to organizations which should be sufficient for the time being.

arve0 commented:
@jjhelmus Thanks for replying back. Appveyor is OK when compiling cython>=0.21 from source. I think binstar will be an improvement as it's https. Also, https should be an requirement if builds/wheels from appveyor are to be uploaded in the future (saw that commented somewhere).

matthew-brett commented:
I may have misunderstood, but I think it's better to make a new virtualenv
that we control, rather than rely on whatever stuff that travis decided to
put into their own virtualenv.

I suppose the pip error is because pip somehow got upgraded, my guess is,
because travis quietly upgraded virtualenv - but it's just a guess.

I think fixing the pip error is just some variant of:

pip install -f travis-wheels.scipy.org --trusted-host
travis-wheels.scipy.org numpy

and I think you will have to do this whether you use travis' virtualenvs,
or make your own.

arve0 commented:
@matthew-brett Thanks for your input. I think you are misunderstanding, as we're not talking about relying on travis virtualenv, rather the python *version* (2.7.9) that ships in their virtualenv. In other words, not using system python (2.7.3). This is only applies to the python 2.7 test, as we never have (as I'm aware of) installed system python 3 and tested with that. The implication is that we cannot use system packages, and if we want to test against PIL, we need to provide a package for that (preferably a wheel package).

Though I understand your point, both @matthew-brett and @blink1073, about not relying on travis env, I must say it's a bit strange. Any package changes in the env that are within our requirements should pass the tests. Also, if installing some non-requirement break the build, I say that's worth looking into. I might be wrong though, I have not investigated your trouble with using travis env earlier.

Regarding `--trusted-host`, that's another issue than the main obstacle here (a `pkg_resources` conflict between system and travis python 2.7). You are correct though, `--trusted-host` is required whenever pip in the venv upgrades to 7.0.

matthew-brett commented:
Thanks for clarifying.

The point about not relying on the travis env, is that we don't have to
track down why things suddenly started to fail for no apparent reason -
where the actual reason is that travis upgraded something.  By using our
own virtualenvs and wheels, we have control over what goes into the tests,
so we are the ones responsible for the breakage and it's easy to see what's
going on.

jjhelmus commented:
FYI, I created the [scikit-image-wheels organization on binstar](https://binstar.org/scikit-image-wheels) and uploaded a number of Gohlke's 32-bit Window wheels to the organization.  I'm happy to authorize others to add packages to the organization as needed.  Binstar can be used as a Python Package Index with pip which can install these wheels.  I created the [install_instructions_jjhelmus branch](https://github.com/jjhelmus/scikit-image/tree/installation_instructions_jjhelmus) which installs the dependencies from binstar.  Total time to install Python, pip, and dependencies is ~2 minutes, ~17 minutes for the full run and [all the tests pass](https://ci.appveyor.com/project/JonathanHelmus/scikit-image/build/1.0.69).  Let me know if there is interest in rolling these changed into this PR.

stefanv commented:
@jjhelmus Is there any advantage to using Binstar rather than our own webserver for hosting the packages?  Also, we should find a way of building the packages ourselves in an open and reproducible way (potentially using wine under Travis CI?)

blink1073 commented:
@stefanv, something like this: https://github.com/cython/cython/wiki/BuildingWindowsInstaller?

blink1073 commented:
@arve0, all of the minimum dependencies (including PIL) should be available now on http://travis-wheels.scikit-image.org.

stefanv commented:
@blink1073 My thought exactly :)  [Literally, if you look at that Wiki page history!]

jjhelmus commented:
@stefanv Binstar implements a PyPI compatible server, meaning that it can respond to pip when it requests a list of packages, versions, metadata, and such.  Basically, binstar acts as a full replacement for PyPI with a much more limited number of packages.  One benefit of this option is that multiple people can upload wheels to the binstar organization and the security (including https), login, hosting, and maintenance is taken care of by binstar.  

Pip can also parse a URL for links to archives which is what the webserver with a directory of whl files option was doing.  This is very simple but does require us or the sklearn/rackspace folks to maintain the web server, manager logins, etc.  Also, I should mention that I do not think the `--trusted-host` switch is required if you use the SSLed version of the rackspace wheelhouse, https://28daf2247a33ed269873-7b1aad3fab3cc330e1fd9d109892382a.ssl.cf2.rackcdn.com/.  

 Truthfully both are good options, neither is technically superior to the other.  I was curious if the binstar option would work so I tried it out, nothing more. The question really comes down to if "we" want to manager a webserver or trust Continuum to do it by using binstar.

arve0 commented:
Regarding win build, doesn't minigw restrict us to 32 bit and 3.6 GB memory only? Win Python 2.7 32 bit halted on me last time I did filtering with large images. Using 64 bit went fine.

arve0 commented:
I'll answer my self, lmgtfy.com/?q=mingw+64, bit is available but seems on need to build it.

arve0 commented:
*crossing fingers*

stefanv commented:
@jjhelmus As long as we host our own packages and rely on no external or hidden builds, I am OK with external hosting infrastructure, given that we can rapidly change deployment to a static server should such a need arise.

Matthew, conveniently, is on my campus and is actively involved in maintaining the infrastructure we use, so I typically just defer to his judgement on this matter.  I believe we share the same philosophy / concerns.

stefanv commented:
It's very possible that the mingw build is unsuitable for 64-bit, and that it causes subtle issues when combined with Python versions using other runtimes.  These are matters currently under deep discussion on the NumPy mailing list.

stefanv commented:
AppVeyor is happy!

stefanv commented on appveyor.yml:
```diff
@@ -37,7 +37,7 @@ install:
 
   # Install the build and runtime dependencies of the project.
-  - "%CMD_IN_ENV% pip install -v %WHEELHOUSE% -r tools/appveyor/requirements.txt"
+  - "%CMD_IN_ENV% pip install -v --trusted-host 28daf2247a33ed269873-7b1aad3fab3cc330e1fd9d109892382a.r6.cf2.rackcdn.com %WHEELHOUSE% -r tools/appveyor/requirements.txt"
```
I presume we can't use travis-wheels.scikit-image.org due to https issues?  But a note here to say that those are the same would be good

jjhelmus replied:
The two sites are not the same, http://28daf2247a33ed269873-7b1aad3fab3cc330e1fd9d109892382a.r6.cf2.rackcdn.com hosts the Windows wheels, http://travis-wheels.scikit-image.org hosts the linux. This could be combined into a single container.

stefanv replied:
Either we can combine them, or I can register the new URL.  Let me know.

arve0 commented:
2.7 will fail..

JDWarner commented:
I was literally seconds away from making a PR that just included the AppVeyor secure link fix. If Travis continues to be an issue, I'll go ahead and submit that one-liner to get the test suite operational.

Travis is behind today, still in queue from over an hour ago...

arve0 commented:
Anyone with access could stop [this test](https://travis-ci.org/scikit-image/scikit-image/builds/64325913), it will most likely fail.

JDWarner commented:
@arve0 Done.

arve0 commented:
> Could not find a version that satisfies the requirement matplotlib==1.1.0 (from versions: 1.1.1, 1.3.1, 1.4.0, 1.4.2)

Bump the mpl dependency or create wheel?

jjhelmus commented:
The [container based infrastructure](http://docs.travis-ci.com/user/workers/container-based-infrastructure/) typically runs faster on Travis, would moving to this be an option?  The big limitation is that you cannot use sudo, but the [apt](http://docs.travis-ci.com/user/apt/) add on can be used to install APT packages, assuming they have been [whitelisted](https://github.com/travis-ci/apt-package-whitelist).

JDWarner commented:
@arve0 While I like backwards compatibility, I don't have a problem with requiring a particular major point release instead of a fully specified string, like 1.1 instead of 1.1.0, as minor point releases tend to be bugfixes and the maintainers don't want to distribute the old, buggy versions. Also, assuming good release behavior, anyone who had 1.1.0 presumably can update to 1.1.1 safely.

So I'd be OK with bumping the requirement to 1.1.1, but @stefanv might disagree.

blink1073 commented:
I'm good with it.  The default matplotlib in 12.04 is 1.1.1rc, which is why we had 1.1.0, but I see no problem with using 1.1.



—
Sent from Mailbox

On Wed, May 27, 2015 at 6:51 PM, Josh Warner <notifications@github.com>
wrote:

> @arve0 While I like backwards compatibility, I don't have a problem with requiring a particular major point release instead of a fully specified string, like 1.1 instead of 1.1.0, as minor point releases tend to be bugfixes and the maintainers don't want to distribute the old, buggy versions. Also, assuming good release behavior, anyone who had 1.1.0 presumably can update to 1.1.1 safely.
> So I'd be OK with bumping the requirement to 1.1.1, but @stefanv might disagree.
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/scikit-image/scikit-image/pull/1526#issuecomment-106112925

arve0 commented:
Assuming `matplotlib==1.1` will resolve to 1.1.1.

JDWarner commented:
@arve0 I think it may need to be `matplotlib>=1.1,<1.1.99`

blink1073 commented:
Perhaps leave requirements.txt alone and explicitly install 1.1.1 from the wheelhouse.

blink1073 commented:
@arve0, I added matplotlib 1.1.0 for completeness sake.

arve0 commented:
Anyone have an idea whats causing 3.2 to [fail](https://travis-ci.org/arve0/scikit-image/jobs/64343285#L6268)?

arve0 commented:
> Could not find a version that satisfies the requirement numpy==1.6.1 (from versions: 1.5.1, 1.6.0, 1.8.1, 1.9.0, 1.9.1, 1.9.2)
  No distributions matching the version for numpy==1.6.1
numpy 1.6.1 is also missing in wheel house.

arve0 commented:
Seems like the 3.2 is due to pillow not closing file pointers. Might be related to this: https://github.com/python-pillow/Pillow/issues/1144

arve0 commented:
2.7 segfaults (numpy 1.6.0 not supported)? https://travis-ci.org/arve0/scikit-image/jobs/64372909

blink1073 commented:
@arve0, I just built numpy 1.6.1 for the wheelhouse.

Try this fix for the intermittent 3.2 error:

```diff
diff --git a/skimage/io/tests/test_pil.py b/skimage/io/tests/test_pil.py
index cd4e893..07e558e 100644
--- a/skimage/io/tests/test_pil.py
+++ b/skimage/io/tests/test_pil.py
@@ -230,9 +230,9 @@ class TestSaveTIF:
     def roundtrip(self, dtype, x, compress):
         with temporary_file(suffix='.tif') as fname:
             if dtype == np.bool:
-                expected = ['low contrast']
+                expected = ['low contrast|unclosed file']
             else:
-                expected = []
+                expected = ['unclosed file|\A\Z']
             with expected_warnings(expected):
                 if compress > 0:
                     imsave(fname, x, compress=compress)
```

blink1073 commented:
@arve0, mind if I stop the previous builds in favor of the latest?

arve0 commented:
@blink1073 Go ahead. Made me think that building all PRs might not be a good idea, but I'm uncertain if its possible to build PRs "on command"?

arve0 commented:
Looks like it's gonna pass now: https://travis-ci.org/arve0/scikit-image/builds/64385969

A though: A binstar organization channel might be a good fit for official windows install. What do you think?

blink1073 commented:
@arve0, I'll let @matthew-brett weight in on binstar vs. rackspace.

blink1073 commented:
I canceled the builds, and yes, they can be re-run at any time.

blink1073 commented:
@arve0, what happened to your updated install instructions?

arve0 commented:
@blink1073, @matthew-brett: Just to be clear and explicit, what I'm thinking of is for *Windows users* of scikit-image (which are not using a python distribution), not the test system. 

`pip install --find-links=binstar-channel scikit-image` is a major improvement to
- download goehlke packages
- install one by one (hope there is no unresolvable deps)
- then install scikit-image

Not sure what happened to the updated install instructions, I'll try to find them.

arve0 commented:
Found the updated install instructions, thank god (Linus) for git. Cleaned up some commits.

arve0 commented:
You may want to stop the test in favor for #1530, as the test has already passed here: https://travis-ci.org/arve0/scikit-image

stefanv commented:
@arve The binstar command looks fine, but would look much better if we can figure out how to build our own windows packages.

This is taking forever to debug; I have a branch on which I am trying out Drone, which seems to be more responsive (at least until they reach the same level of success as Travis-CI, I presume)

blink1073 commented:
@stefanv, I love everything about drone.io except for not having parallel builds.

blink1073 commented:
@stefanv, we could be a bit smarter about our builds.  We could do one "extra" thing per build: docs, gallery, app gallery, optional dependecies.  






—
Sent from Mailbox

On Thu, May 28, 2015 at 5:15 PM, Stefan van der Walt
<notifications@github.com> wrote:

> @arve The binstar command looks fine, but would look much better if we can figure out how to build our own windows packages.
> This is taking forever to debug; I have a branch on which I am trying out Drone, which seems to be more responsive (at least until they reach the same level of success as Travis-CI, I presume)
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/scikit-image/scikit-image/pull/1526#issuecomment-106615520

JDWarner commented:
Thanks for sticking this out, @arve0!

I'm merging this, as it is blocking the dask[array] parallel PR and should fix the intermittent unclosed file error. Any other tweaks, improvements, etc. can happen in separate PRs.

blink1073 commented:
Well done @arve0, I owe you a beverage of choice!

arve0 commented:
More likely I owe you guys a beverage of choice for keeping scikit-image up to such a high quality!

stefanv commented on doc/source/install.txt:
```diff
@@ -1,53 +1,153 @@
-Pre-built installation
+Installing scikit-image
+-----------------------
+If you are on Mac OS X you're lucky, open the terminal and install
```
Why the comment about people on OSX being lucky?  I'd get rid of this.

matthew-brett replied:
Surely you did not forget that OSX is the world's most advanced operating system?

And that the latest MacBook is also available in gold?

arve0 replied:
Because wheels exists and installation is straight forward containing only one step.

matthew-brett replied:
But joking aside, in this particular case, if you are on OSX, you are lucky.   Not a value judgement about operating systems, just a comment on the convenience of pip installs.

stefanv replied:
It's strange to find something funny and so disturbing at the same time.

stefanv commented on doc/source/install.txt:
```diff
@@ -1,53 +1,153 @@
+
+Linux, Mac and Windows
 ----------------------
+An easy light weight method to get scikit-image installed on all of the most
```
These instructions are insufficient.  I will rework.

arve0 replied:
What are they missing?

stefanv replied:
I'll have a PR ready in a few minutes.

stefanv commented:
@JDWarner This was a premature merge on the docs front.

JDWarner commented:
@stefanv  I was going to address that in a separate PR as the technical work was done here, blocking other PRs, and discussions about wording can get lengthy.

Didn't see your comments until I submitted #1533 just now.

stefanv commented:
See https://github.com/scikit-image/scikit-image/pull/1535
